File: Natural_Language_Processing/Section 36 - Natural Language Processing/nlp_classifier.py
'''
Create a Bog of Words model.
Run the model through multiple classification algorithms.
Select the best classification algorithm.
'''
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer

# Print breaker
def breaker():
    print("="*50)

# Print results
def print_results(cm, accuracy, error_rate, percision, recall, f1, source):
    print("Results for: {}".format(source))
    print("Confusion matrix: \n{}".format(cm))
    print("Accuracy: {}".format(accuracy))
    print("Error Rate: {}".format(error_rate))
    print("Percision: {}".format(percision))
    print("Recall: {}".format(recall))
    print("F1: {}".format(f1))

# Clean data prior to classifying'
def data_cleaner(data):
    # Clean the text to prepare for models being built
    # We will remove upper case + unneeded puncutation and unneeded words (the, them, an...)
    corpus = []
    ps = PorterStemmer()
    for i in range(0,len(data)):
        review = re.sub('[^a-zA-Z]', ' ', data['Review'][i])                                    # Remove non-alphabetical chars
        review = review.lower()                                                                    # Set all characters to lowercase
        review = review.split()                                                                    # Split the string into character array
        review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))] # Remove common words and stem the words into their root ex: loved => love
        review = ' '.join(review)                                                                  # Join the strings back together into one
        corpus.append(review)

    return(corpus)

# Create Bag of Words model
def bag_of_words(corpus):
    # Creating the Bag of Words (BoW) model
    # Creating datatable of words existing in each review (sparse matrix)
    cv = CountVectorizer(max_features = 1500) # Max features leads to less sparcity in the matrix
    X = cv.fit_transform(corpus).toarray()
    return(X)

# Confusion Matrix
def confusion_matrix(y_test, y_pred):
    # Making the Confusion Matrix
    from sklearn.metrics import confusion_matrix
    cm = confusion_matrix(y_test, y_pred)
    tn = cm[0][0]
    tp = cm[1][1]
    fn = cm[0][1]
    fp = cm[1][0]

    accuracy = (tp+tn) / (tn+tp+fp+fn)
    error_rate = 1 - accuracy
    percision = tp / (tp+fp) if tp>0 or fp>0 else 0.0
    recall = tp / (tp+fn) if tp>0 or fn>0 else 0.0
    f1 = 2*(percision*recall)/(percision+recall) if percision>0 or recall>0 else 0.0
    return(cm, accuracy, error_rate, percision, recall, f1)

# Test the classification of each algorithm
def classify(X, y):
    functions = [random_selection, log_regression, naive_bayes, knn, svm, kernel_svm, decision_trees, rand_forest, cart]
    best_f1 = 0
    best_classifier = ""
    # Splitting the Bag of Words into the Training set and Test set
    from sklearn.cross_validation import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)
    breaker() # Initial Break

    for func in functions:
        class_pred, class_source = func(X_train, y_train, X_test, y_test)
        class_cm, class_accuracy, class_error_rate, class_percision, class_recall, class_f1 = confusion_matrix(y_test, class_pred)
        print_results(cm = class_cm, accuracy = class_accuracy, error_rate = class_error_rate, percision = class_percision, recall = class_recall, f1 = class_f1, source = class_source)
        if class_f1 > best_f1:
            best_f1 = class_f1
            best_classifier = class_source
        breaker()

    print("Best classifier: {}\nF1 Score: {}".format(best_classifier, best_f1))

# Define each algorithm
def random_selection(X_train, y_train, X_test, y_test):
    import random
    y_pred = []
    for i in range(0,len(y_test)):
        random_select = 0 if random.random() < 0.5 else 1
        y_pred.append(random_select)

    return(y_pred, "Random")

def naive_bayes(X_train, y_train, X_test, y_test):
    # Fitting Naive Bayes to the Training set
    from sklearn.naive_bayes import GaussianNB
    classifier = GaussianNB()
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Naive Bayes")

def log_regression(X_train, y_train, X_test, y_test):
    # Fitting Logistic Regression to the Training set
    from sklearn.linear_model import LogisticRegression
    classifier = LogisticRegression()
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Logistic Regression")

def knn(X_train, y_train, X_test, y_test):
    # Fitting K-NN to the Training set
    from sklearn.neighbors import KNeighborsClassifier
    classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "K Nearest Neighbor")

def svm(X_train, y_train, X_test, y_test):
    # Fitting SVM to the Training set
    from sklearn.svm import SVC
    classifier = SVC(kernel = 'linear', random_state = 0)
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Linear SVM")

def kernel_svm(X_train, y_train, X_test, y_test):
    # Fitting Kernel SVM to the Training set
    from sklearn.svm import SVC
    classifier = SVC(kernel = 'rbf', random_state = 0)
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Kernel SVM")

def decision_trees(X_train, y_train, X_test, y_test):
    # Fitting Decision Tree Classification to the Training set
    from sklearn.tree import DecisionTreeClassifier
    classifier = DecisionTreeClassifier(criterion = 'entropy')
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Decision Trees")

def rand_forest(X_train, y_train, X_test, y_test):
    # Fitting Random Forest Classification to the Training set
    from sklearn.ensemble import RandomForestClassifier
    classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "Random Forest")

def cart(X_train, y_train, X_test, y_test):
    # Fitting Decision Tree Classification to the Training set
    from sklearn.tree import DecisionTreeClassifier
    classifier = DecisionTreeClassifier()
    classifier.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(X_test)
    return(y_pred, "CART")

def maxent(X_train, y_train, X_test, y_test):
    algorithm = nltk.classify.MaxentClassifier.ALGORITHMS[0]
    classifier = nltk.MaxentClassifier.train(X_train, algorithm)
    classifier.show_most_informative_features(10)

    y_pred = []
    for row in X_test:
        determined_label = classifier.classify(row)
        y_pred.append(determined_label)

    return(y_pred, "Max Entropy")

if __name__ == "__main__":
    # Import the data set
    dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

    # Clean data
    corpus = data_cleaner(dataset)

    # Create bag of words model
    features = bag_of_words(corpus)
    target = dataset.iloc[:,1]

    # Create and print results of classification models
    classify(X = features, y = target)

File: Natural_Language_Processing/Section 36 - Natural Language Processing/nlp_classifier.R
# NLP Classifiers

# Import Libraries
library(tm) # Used for sparce matrix manipulation
library(SnowballC) # Used for stopwords
library(caTools)
library(randomForest)
library(rpart)
library(e1071)
library(class)
library(C50)
library(maxent)

getCorpus <- function(data){
  corpus = VCorpus(VectorSource(data$Review)) # Create corpus
  corpus = tm_map(corpus, content_transformer(tolower)) # Convert all letters to lowercase
  corpus = tm_map(corpus, removeNumbers) # Remove any numbers
  corpus = tm_map(corpus, removePunctuation) # Remove any punctuations
  corpus = tm_map(corpus, removeWords, stopwords()) # Remove unneeded words (and, this, the etc.)
  corpus = tm_map(corpus, stemDocument) # take just root of each word
  corpus = tm_map(corpus, stripWhitespace) # Remove whitespace

  return(corpus)
}

bagOfWords <- function(corpus){
  # Create the Bag of Words model
  dtm = DocumentTermMatrix(corpus)
  dtm = removeSparseTerms(dtm, 0.999)
  dataset = as.data.frame(as.matrix(dtm))

  return(dataset)
}

breaker <- function() {
  cat(paste0(rep("=", 50), sep = ""))
  cat(paste0("\n"))
}

confusionMatrix <- function(y_pred, y_actual){
  # Making the Confusion Matrix
  cm = table(y_actual, y_pred)
  tn = cm[1]
  fp = cm[2]
  fn = cm[3]
  tp = cm[4]

  # Calculate other metrics
  accuracy = (tp+tn) / (tn+tp+fp+fn)
  error_rate = 1 - accuracy
  if(tp>0 || fp>0) percision = tp / (tp+fp)  else 0.0
  if(tp>0 || fn>0) recall = tp / (tp+fn) else 0.0
  if(percision>0 || recall>0) f1 = 2*(percision*recall)/(percision+recall) else 0.0

  return(list(cm, accuracy, error_rate, percision, recall, f1))
}

printResults <- function(results, source){

  # Print results
  cat(paste("Results for", source, "\n"))
  cat(paste("Confusion Matrix:", results[1], "\n"))
  cat(paste("Accuracy:", results[2], "\n"))
  cat(paste("Error Rate:", results[3], "\n"))
  cat(paste("Percision:", results[4], "\n"))
  cat(paste("Recall:", results[5], "\n"))
  cat(paste("F1 Score:", results[6], "\n"))
  breaker()
}

classifyDataset <- function(dataset){
  # Splitting the dataset into the Training set and Test set
  # install.packages('caTools')
  library(caTools)
  set.seed(123)
  split = sample.split(dataset$Liked, SplitRatio = 0.8)
  training_set = subset(dataset, split == TRUE)
  test_set = subset(dataset, split == FALSE)
  y_actual = test_set[, 692]

  # Initial line break
  breaker()

  # Define funcion list
  func_list <- list(randForest,
                    decsionTrees,
                    naiveBayesClassifier,
                    kernelSVM,
                    linearSVM,
                    knnClassifier,
                    logRegression,
                    cFiveZero,
                    maxEntropy)

  # Execute each function
  best_class = ""
  best_score = 0
  for(func in func_list){
    func_results = func(training_set, test_set)
    cm_results = confusionMatrix(func_results$predictions, y_actual)
    # cat(as.character(cm_results))
    func_f1 = as.numeric(cm_results[6])
    if(func_f1 > best_score){
      best_score = func_f1
      best_class = func_results$source
    }
    printResults(cm_results, func_results$source)
  }

  breaker()
  cat(paste("\nBest Results:\n", "Class:", best_class, "\nScore:", best_score, "\n"))
}

randForest <- function(training_set, test_set){
  # Fitting Random Forest Classification to the Training set
  classifier = randomForest(x = training_set[-692], y = training_set$Liked, ntree = 10)

  # Predicting the Test set results
  y_pred = predict(classifier, newdata = test_set[-692])
  return(list("predictions" = y_pred, "source" = "Random Forest"))
}

decsionTrees <- function(training_set, test_set){
  # Fitting Decision Tree Classification to the Training set
  classifier = rpart(formula = Liked ~ .,
                     data = training_set)

  # Predicting the Test set results
  y_pred = predict(classifier, newdata = test_set[-692], type = 'class')
  return(list("predictions" = y_pred, "source" = "Decision Trees"))
}

naiveBayesClassifier <- function(training_set, test_set){
  classifier = naiveBayes(x = training_set[-692], y = training_set$Liked)

  # Predicting the Test set results
  y_pred = predict(classifier, newdata = test_set[-692])
  return(list("predictions" = y_pred, "source" = "Naive Bayes"))
}

kernelSVM <- function(training_set, test_set){
  classifier = svm(formula = Liked ~ .,
                   data = training_set,
                   type = 'C-classification',
                   kernel = 'radial')

  # Predicting the Test set results
  y_pred = predict(classifier, newdata = test_set[-692])
  return(list("predictions" = y_pred, "source" = "Kernel SVM"))
}

linearSVM <- function(training_set, test_set){
  classifier = svm(formula = Liked ~ .,
                   data = training_set,
                   type = 'C-classification',
                   kernel = 'linear')

  # Predicting the Test set results
  y_pred = predict(classifier, newdata = test_set[-692])
  return(list("predictions" = y_pred, "source" = "Linear SVM"))
}

knnClassifier <- function(training_set, test_set){
  y_pred = knn(train = training_set[, -692],
               test = test_set[, -692],
               cl = training_set[, 692],
               k = 5,
               prob = TRUE)
  return(list("predictions" = y_pred, "source" = "k Nearest Neighbor"))
}

logRegression <- function(training_set, test_set){
  classifier = glm(formula = Liked ~ .,
                   family = binomial,
                   data = training_set)

  # Predicting the Test set results
  prob_pred = predict(classifier, type = 'response', newdata = test_set[-692])
  y_pred = ifelse(prob_pred > 0.5, 1, 0)
  return(list("predictions" = y_pred, "source" = "Logistic regression"))
}

cFiveZero <- function(training_set, test_set){
  classifier = C5.0(formula = Liked ~ .,
                    data = training_set,
                    rules = TRUE)
  y_pred = predict(classifier, newdata = test_set[-692])
  return(list("predictions" = y_pred, "source" = "C5.0"))
}

maxEntropy <- function(training_set, test_set){
  classifier = maxent(training_set[,-692], training_set[,692])

  y_pred = predict(classifier, test_set[,-692])
  y_pred = y_pred[, "labels"]
  return(list("predictions" = y_pred, "source" = "Maximum Entropy"))
}

main <- function() {
  # Import the dataset
  source("/Users/rheft/dev/Sonny/config.R")
  dataset_og = read.delim(paste0(data_locale,'Restaurant_Reviews.tsv'), quote='')

  corpus = getCorpus(dataset_og)
  bow_data = bagOfWords(corpus)
  bow_data$Liked = dataset_og$Liked

  # Encoding the target feature as factor
  bow_data$Liked = factor(bow_data$Liked, levels = c(0, 1))

  # Classify the dataset using each classifier
  classifyDataset(bow_data)
}

main()
